### PR TITLE
Work on tests

### PR DESCRIPTION
**Pull request by gytis-ivaskevicius**
I started writing somewhat proper tests due to /examples/* becoming unmanageable and the overall lack of test-cases we are actually covering.

### CODE REVIEW

**Comment by gytis-ivaskevicius** [comment]
Please double-check if I forgot some important test cases.
Did not test modules flow since it is so simple 
Did not test exporters since I haven't looked into em and the one for packages seems to be broken? :thinking:  (examples/exporters)

Will add more examples and probably few more tests in the future

**Comment by gytis-ivaskevicius** [comment]
Also, I'd like to dump pretty much everything to `/src/` or `/nix/` folder (haven't decided on the name) to the cleanup root directory as well as rename exporters.
For example `utils.lib.exporter.overlaysFromChannelsExporter` - mentioning `exporter` twice is definetelly unecessary. 

Anyways, what kind of naming would you guys like?

**Comment by blaggacao** [comment]
> Anyways, what kind of naming would you guys like?

Agree double naming is inappropriate. First there was the name, then there was the namespace, now the name needs renaming.

**Review by blaggacao — APPROVED** [review]
LGTM. This is a really compounding stabilizing measure!

File: devshell.toml [hunk 1]
```diff
@@ -5,70 +5,53 @@ packages = [
   "nixpkgs-fmt",
```
[the remark this answers is not in the document]
Comment (gytis-ivaskevicius) [reply devshell.toml:5]: Yeah, i forgot about it. Will do!

File: shell.nix [hunk 2]
```diff
@@ -17,5 +17,53 @@ let
...
+  test = name: withCategory "tests" {
+    name = "check-${name}";
+    help = "Checks ${name} testcases";
+    command = "cd $DEVSHELL_ROOT/tests/${name} && nix flake show && nix flake check";
+  };
+
+  dry-nixos-build = example: host: withCategory "dry-build" {
+    name = "build-${example}-${host}";
+    command = "cd $DEVSHELL_ROOT/examples/${example} && nix flake show && nix build .#nixosConfigurations.${host}.config.system.build.toplevel --dry-run";
+  };
```
Comment (Pacman99) [inline shell.nix:32]: I don't know if this would be necessary but the commands could be prepended with `nix flake lock --update-input utils`

**Review by Pacman99 — APPROVED** [review]
This looks amazing!

Regarding renames, I would like to switch `exporter` to `exporters` and `builder` to `builders`. Thats how nixpkgs lib goes, `lib.lists`, `lib.strings`. So the non-plural is confusing to me.

Also for examples/tests, the `.gitignore` line I added doesn't seem to be working. It would be nice to address that. Having to delete and recreate lock files is very annoying.

**Comment by blaggacao** [comment]
> Also for examples/tests, the .gitignore line I added doesn't seem to be working. It would be nice to address that. Having to delete and recreate lock files is very annoying.

Isn't this because of `nix flake` command force-adding them?

File: examples/exporters/flake.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+
+      # export overlays automatically for all packages defined in overlaysBuilder of each channel
+      overlays = utils.lib.exporter.overlaysFromChannelsExporter {
+        inherit (self) pkgs inputs;
+      };
+
+      # TODO: Broken
+      # construct packagesBuilder to export all packages defined in overlays
+      #packagesBuilder = utils.lib.builder.packagesFromOverlaysBuilderConstructor self.overlays;
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply examples/exporters/flake.nix:28]: `utils.lib.builder."pkgs<-overlays"`

**Pull request merged**